crates/bevy_math/src/bounding/bounded2d.rs

```diff
@@ -0,0 +1,331 @@
+            max: Vec2::new(1., 0.5),
+        };
+        let b = Aabb2d {
+            min: Vec2::new(-2., -0.5),
+            max: Vec2::new(0.75, 1.),
+        };
+        let merged = a.merged(&b);
+        assert!((merged.min - Vec2::new(-2., -1.)).length() < std::f32::EPSILON);
+        assert!((merged.max - Vec2::new(1., 1.)).length() < std::f32::EPSILON);
+        assert!(merged.contains(&a));
+        assert!(merged.contains(&b));
+        assert!(!a.contains(&merged));
+        assert!(!b.contains(&merged));
+    }
+
+    #[test]
+    fn padded() {
+        let a = Aabb2d {
+            min: Vec2::new(-1., -1.),
+            max: Vec2::new(1., 1.),
+        };
+        let padded = a.padded((Vec2::ONE, Vec2::Y));
+        assert!((padded.min - Vec2::new(-2., -2.)).length() < std::f32::EPSILON);
+        assert!((padded.max - Vec2::new(1., 2.)).length() < std::f32::EPSILON);
+        assert!(padded.contains(&a));
+        assert!(!a.contains(&padded));
+    }
+
+    #[test]
+    fn shrunk() {
+        let a = Aabb2d {
+            min: Vec2::new(-1., -1.),
+            max: Vec2::new(1., 1.),
+        };
+        let shrunk = a.shrunk((Vec2::ONE, Vec2::Y));
+        assert!((shrunk.min - Vec2::new(-0., -0.)).length() < std::f32::EPSILON);
+        assert!((shrunk.max - Vec2::new(1., 0.)).length() < std::f32::EPSILON);
+        assert!(a.contains(&shrunk));
+        assert!(!shrunk.contains(&a));
+    }
+}
+
+use crate::primitives::Circle;
+
+/// A bounding circle
+#[derive(Clone)]
+pub struct BoundingCircle {
+    /// The center of the bounding circle
+    pub center: Vec2,
+    /// The circle
+    pub circle: Circle,
+}
+
+impl BoundingCircle {
+    /// Construct a bounding circle from its center and radius
+    #[inline(always)]
+    pub fn new(center: Vec2, radius: f32) -> Self {
+        debug_assert!(radius >= 0.);
+        Self {
+            center,
+            circle: Circle { radius },
+        }
+    }
+
+    /// Get the radius of the bounding circle
+    #[inline(always)]
+    pub fn radius(&self) -> f32 {
+        self.circle.radius
+    }
+}
+
+impl BoundingVolume for BoundingCircle {
+    type Position = Vec2;
+    type Padding = f32;
+
+    #[inline(always)]
+    fn center(&self) -> Self::Position {
+        self.center
+    }
+
+    #[inline(always)]
+    fn visible_area(&self) -> f32 {
+        std::f32::consts::PI * self.radius() * self.radius()
+    }
+
+    #[inline(always)]
+    fn contains(&self, other: &Self) -> bool {
+        let furthest_point = (other.center - self.center).length() + other.radius();
+        furthest_point <= self.radius()
+    }
+
+    #[inline(always)]
+    fn merged(&self, other: &Self) -> Self {
+        let diff = other.center - self.center;
+        let length = diff.length();
+        if self.radius() >= length + other.radius() {
+            return self.clone();
+        }
+        if other.radius() >= length + self.radius() {
+            return other.clone();
+        }
```

Comment:
we could probably defer the `sqrt` from `length` to after the early returns here and use the same logic as above to save some `sqrt` ops.

Reply:
I think this could be a good optimisation, but I feel it does hinder code-readability so I think a benchmark to test whether it makes a positive difference might be desirable here.

Reply:
Yeah, this shouldn't be a blocker. It would be nice to put a comment in the code if it isn't included though. Just so we don't forget about it.

Reply:
The code looks a bit worse, but overall not too bad.
```diff
     #[inline(always)]
     fn merged(&self, other: &Self) -> Self {
         let diff = other.center - self.center;
+
-        let length = diff.length();
+        // We use length squared since it's cheaper in the early return cases
+        let length_squared = diff.length_squared();
+        let radius_diff = self.radius() - other.radius();
+        let radius_diff_squared = radius_diff.powi(2).copysign(radius_diff);
-        if self.radius() >= length + other.radius() {
+        if radius_diff_squared >= length_squared {
             return self.clone();
         }
-        if other.radius() >= length + self.radius() {
+        if -radius_diff_squared >= length_squared {
             return other.clone();
         }
+        let length = length_squared.sqrt();
         let dir = diff / length;
         Self::new(
-            (self.center + other.center) / 2. + dir * ((other.radius() - self.radius()) / 2.),
+            (self.center + other.center) / 2. + dir * radius_diff / -2.,
             (length + self.radius() + other.radius()) / 2.,
         )
     }
```
Possibly worse is that doing it this would make the case where neither `self` nor `other` contains the other slower.
Either way for this change I think we should probably make benchmarks first. Benchmarks seem like a reasonable thing to add to something as performance critical as bounding volume logic.